AdafruitMotorHat/lib/src/main/java/com/zugaldia/adafruit/motorhat/library/AdafruitStepperMotor.java

```diff
@@ -0,0 +1,441 @@
+        }
+
+        // Go to next 'step' and wrap around
+        currentstep += (MICROSTEPS * 4);
+        currentstep %= (MICROSTEPS * 4);
+
+        // Only really used for microstepping, otherwise always on!
+        if (!alreadySetPWM || style == AdafruitMotorHat.MICROSTEP) {
+
+            // If we're not doing microstepping, then the value will always be the same,
+            // so only do it once.  Otherwise it wastes 5 milliseconds according to logs.
+            alreadySetPWM = true;
+            MC.getPwm().setPWM(PWMA, 0, pwm_a * 16);
+            MC.getPwm().setPWM(PWMB, 0, pwm_b * 16);
+        }
+
+        if (style == AdafruitMotorHat.MICROSTEP) {
+            if (currentstep >= 0 && currentstep < MICROSTEPS) {
+                coils = new int[]{1, 1, 0, 0};
+            } else if (currentstep >= MICROSTEPS && currentstep < MICROSTEPS * 2) {
+                coils = new int[]{0, 1, 1, 0};
+            } else if (currentstep >= MICROSTEPS * 2 && currentstep < MICROSTEPS * 3) {
+                coils = new int[]{0, 0, 1, 1};
+            } else if (currentstep >= MICROSTEPS * 3 && currentstep < MICROSTEPS * 4) {
+                coils = new int[]{1, 0, 0, 1};
+            }
+        } else {
+            coils = step2coils[currentstep / (MICROSTEPS / 2)];
+        }
+
+        //Log.d(".", "coils state = " + coils);
+
+        MC.setPin(AIN2, coils[0]);
+        MC.setPin(BIN1, coils[1]);
+        MC.setPin(AIN1, coils[2]);
+        MC.setPin(BIN2, coils[3]);
+
+        return currentstep;
+    }
+
+    public void step(int steps, int direction, int stepstyle) {
+
+        double s_per_s = sec_per_step;
+        int lateststep = 0;
+
+        if (stepstyle == AdafruitMotorHat.INTERLEAVE) {
+            s_per_s /= 2.0;
+        } else if (stepstyle == AdafruitMotorHat.MICROSTEP) {
+            s_per_s /= MICROSTEPS;
+            steps *= MICROSTEPS;
+        }
+
+        Log.d(".", "secs per step: " + s_per_s);
+
+        for (int s = 1; s <= steps; s++) {
+
+            lateststep = oneStep(direction, stepstyle);
+
+            if (sleepBetweenSteps) {
+                try {
+                    Thread.sleep((long) s_per_s * 1000);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+
+        if (stepstyle == AdafruitMotorHat.MICROSTEP) {
+
+            // This is an edge case, if we are in between full steps, lets just keep going
+            // So we end on a full step
+            while (lateststep != 0 && lateststep != MICROSTEPS) {
+
+                lateststep = oneStep(direction, stepstyle);
+
+                if (sleepBetweenSteps) {
+                    try {
+                        Thread.sleep((long) s_per_s * 1000);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+
+
+
+
+/*
+#!/usr/bin/python
```

Comment:
This is looking great - thanks for the contribution. Could you remove this Python code from here before we merge?